gateways/firstdata/firstdata.go

```diff
@@ -0,0 +1,257 @@
+		Response:             firstdataResponse.TransactionState,
+		AvsResultRaw:         avsRawString,
+		CvvResultRaw:         firstdataResponse.SecurityCodeResponse,
+	}, nil
+}
+
+func (client *FirstdataClient) Capture(request *sleet.CaptureRequest) (*sleet.CaptureResponse, error) {
+	firstdataCaptureRequest, err := buildCaptureRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	url := "https://" + client.host + authPath + "/" + request.TransactionReference
+
+	firstdataResponse, err := client.sendRequest(*request.ClientTransactionReference, url, *firstdataCaptureRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.CaptureResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+
+	return &sleet.CaptureResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+// Void cancels a Firstdata payment. If successful, the void response will be returned. A previously voided
+// payment or one that has already been settled cannot be voided.
+func (client *FirstdataClient) Void(request *sleet.VoidRequest) (*sleet.VoidResponse, error) {
+	firstdataVoidRequest, err := buildVoidRequest(request)
+	if err != nil {
+		return nil, err
+	}
+	url := "https://" + client.host + authPath + "/" + request.TransactionReference
+	firstdataResponse, err := client.sendRequest(*request.ClientTransactionReference, url, *firstdataVoidRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.VoidResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+	return &sleet.VoidResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+// Refund refunds a Firstdata payment. If successful, the refund response will be returned. Multiple
+// refunds can be made on the same payment, but the total amount refunded should not exceed the payment total.
+func (client *FirstdataClient) Refund(request *sleet.RefundRequest) (*sleet.RefundResponse, error) {
+	firstdataRefundRequest, err := buildRefundRequest(request)
+	if err != nil {
+		return nil, err
+	}
+	url := "https://" + client.host + authPath + "/" + request.TransactionReference
+	firstdataResponse, err := client.sendRequest(*request.ClientTransactionReference, url, *firstdataRefundRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.RefundResponse{Success: false, ErrorCode: &firstdataResponse.Error.Code}
+		return &response, nil
+	}
+	return &sleet.RefundResponse{Success: true, TransactionReference: firstdataResponse.IPGTransactionId}, nil
+}
+
+// reqId is our internally generated id,unique per request, tranasactionRef is firstdata's returned ref
+func (client *FirstdataClient) TransactionInquiry(reqId, transactionRef string) (*Response, error) {
```

Comment:
what is this being used for?

Reply:
I used it in the integration tests to check the state of a transaction after secondary transactions. So after calling Capture i use inquiry to check if the state is "CAPTURED". It may seem unnecessary but the call to Capture does not actually return the state "CAPTURED" in its response, the inquiry does. 

Also, we can't assume that just because the call to Capture was successful, that the state must be "CAPTURED" because if the endpoint is called with a previously used client-request-id, it will automatically return the original response associated with that request id. So it is possible to get a successful http response from calling capture, but not actually capture the transaction.

Reply:
If its just used for integration tests can we move the function into the test file? I'd like to kee the actual gateway file just for like production processing, if that makes sense?

Reply:
Sure thing